File: apps/backend/schemas/experiments.py
"""Pydantic schemas for experiment workspaces."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


ExperimentStatus = Literal["active", "completed", "archived"]
ExperimentEntryKind = Literal["observation", "file", "metric", "conclusion"]


class ExperimentCreate(BaseModel):
    title: str = Field(min_length=1, max_length=120)
    description: Optional[str] = Field(default=None, max_length=2000)


class ExperimentUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=120)
    description: Optional[str] = Field(default=None, max_length=2000)
    status: Optional[ExperimentStatus] = None


class ExperimentEntryCreate(BaseModel):
    kind: ExperimentEntryKind
    title: str = Field(min_length=1, max_length=160)
    content: Optional[str] = Field(default=None, max_length=20_000)
    metadata: Dict[str, Any] = Field(default_factory=dict)


class ExperimentEntryRead(BaseModel):
    id: str
    experiment_id: str
    kind: ExperimentEntryKind
    title: str
    content: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ExperimentThreadRead(BaseModel):
    id: str
    title: Optional[str] = None
    first_message: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ExperimentRead(BaseModel):
    id: str
    title: str
    description: Optional[str] = None
    status: ExperimentStatus
    thread_count: int = 0
    entry_count: int = 0
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ExperimentListResponse(BaseModel):
    items: List[ExperimentRead] = Field(default_factory=list)


class ExperimentDetailRead(ExperimentRead):
    threads: List[ExperimentThreadRead] = Field(default_factory=list)
    entries: List[ExperimentEntryRead] = Field(default_factory=list)


class ExperimentThreadCreate(BaseModel):
    title: Optional[str] = Field(default=None, max_length=120)

File: apps/dashboard/lib/experiments.ts
export type ExperimentStatus = 'active' | 'completed' | 'archived';
export type ExperimentEntryKind = 'observation' | 'file' | 'metric' | 'conclusion';

export type ExperimentSummary = {
  id: string;
  title: string;
  description: string | null;
  status: ExperimentStatus;
  thread_count: number;
  entry_count: number;
  created_at: string | null;
  updated_at: string | null;
};

export type ExperimentThread = {
  id: string;
  title: string | null;
  first_message: string | null;
  created_at: string | null;
  updated_at: string | null;
};

export type ExperimentEntry = {
  id: string;
  experiment_id: string;
  kind: ExperimentEntryKind;
  title: string;
  content: string | null;
  metadata: Record<string, unknown>;
  created_at: string | null;
  updated_at: string | null;
};

export type ExperimentDetail = ExperimentSummary & {
  threads: ExperimentThread[];
  entries: ExperimentEntry[];
};

async function experimentFetch<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(path, {
    cache: 'no-store',
    ...init,
    headers: {
      'Content-Type': 'application/json',
      ...init?.headers,
    },
  });
  if (!response.ok) {
    const payload = await response.json().catch(() => null) as { detail?: string } | null;
    throw new Error(payload?.detail || `Experiment request failed (${response.status})`);
  }
  if (response.status === 204) return undefined as T;
  return response.json() as Promise<T>;
}

export async function listExperiments(limit = 20): Promise<ExperimentSummary[]> {
  const data = await experimentFetch<{ items: ExperimentSummary[] }>(`/api/experiments?limit=${limit}`);
  return data.items || [];
}

export function getExperiment(experimentId: string): Promise<ExperimentDetail> {
  return experimentFetch(`/api/experiments/${experimentId}`);
}

export function createExperiment(input: { title: string; description?: string }): Promise<ExperimentSummary> {
  return experimentFetch('/api/experiments', { method: 'POST', body: JSON.stringify(input) });
}

export function updateExperiment(
  experimentId: string,
  input: Partial<Pick<ExperimentSummary, 'title' | 'description' | 'status'>>,
): Promise<ExperimentDetail> {
  return experimentFetch(`/api/experiments/${experimentId}`, { method: 'PATCH', body: JSON.stringify(input) });
}

export function createExperimentThread(experimentId: string): Promise<ExperimentThread> {
  return experimentFetch(`/api/experiments/${experimentId}/threads`, {
    method: 'POST',
    body: JSON.stringify({}),
  });
}

export function createExperimentEntry(
  experimentId: string,
  input: { kind: ExperimentEntryKind; title: string; content?: string },
): Promise<ExperimentEntry> {
  return experimentFetch(`/api/experiments/${experimentId}/entries`, {
    method: 'POST',
    body: JSON.stringify(input),
  });
}

export function deleteExperimentEntry(experimentId: string, entryId: string): Promise<void> {
  return experimentFetch(`/api/experiments/${experimentId}/entries/${entryId}`, { method: 'DELETE' });
}
